### DISCUSSION TITLE
Shift+Fn keycodes are not what Midnight Commander expects

### DISCUSSION THREAD

**Discussion by edrozenberg** [ghostty-org/ghostty#3777] [Feature Requests, Ideas] [answered]
Ghostty Version 1.0.0
Build 8469
Commit 4b4d4062
(MacOS)

I haven't been able to find an accepted defined standard for what keycodes should be sent by Shift+Fn combos, maybe there is no standard. If there is no standard then it's up to the user to make any desired changes in Terminal.app, Ghostty, etc.

In the case of Midnight Commander (`mc`), it expects the following:

Create new file:
Shift+F4: `\033[26~` ( i.e. `ESC [26~` )
Ghostty sends: `ESC [1;2S`

Copy to current location:
Shift+F5: `\033[28~` ( i.e. `ESC [28~` )
Ghostty sends: `ESC 15;2~`

Rename current item
Shift+F6: `\033[29~` ( i.e. `ESC [29~` )
Ghostty sends: `ESC [17;2~`

Apple's Terminal.app is more similar to what Midnight Commander expects, but the codes are off by a bit, so I wind up modifying them in Terminal.app terminal key code prefs.

For ex. Shift-F5 sends `[25~` instead of `[28~` by default in Terminal.app:

![Apple Defaults Terminal](https://github.com/user-attachments/assets/4be1927e-9f48-4f0f-9cce-94c097fb9339)

So maybe there is no standard.

**Comment by mitchellh** [comment, marked as answer]
https://www.leonerd.org.uk/hacks/fixterms/

You can remap them manually with `keybind = whatever=text:`

**Comment by edrozenberg** [comment]
Thanks, I was able to duplicate what Ghostty does ootb with:
`keybind = shift+f6=esc:[17;2~` ✅ (works as expected)

But using the correct code `mc` expects for Shift-F6 does not work, `mc` does not respond with a rename dialog.
`keybind = shift+f6=esc:[29~` ❌ (does not work as expected)

**Comment by lpetrus** [comment]
I think a more correct workaround is to set the TERM environment variable.

By default ghostty sets the TERM as "xterm-ghostty".  When this value is set, then mc handles shift + f* keys incorrectly.

However when I set TERM to "xterm-256color" mc started to behave correctly with shfit + f* keys.

**Reply by edrozenberg** [reply]
> when I set TERM to "xterm-256color" mc started to behave correctly with shfit + f* keys.

That's a great find, when setting that `TERM` value the Shift - f4/f5/f6 key combos work perfectly, without even needing any `keybind` reconfig in the Ghostty config file.

Hopefully Ghostty 1.1 will bring some fixes to this somehow, they are working on various input issues for that release, because changing the `TERM` value is not a longterm solution and not recommended as some features Ghostty provides are exposed by its `xterm-ghostty` terminfo definitions.

**Reply by edrozenberg** [reply]
This is still broken on latest `tip` (commit a3e462bb build 9336) on Mac (can't use Shift + F4/F5/F6 in `mc` to perform the expected commands), I'll keep tracking this issue and open a separate discussion later if I track down anything further.
